### PR TITLE
doc: [mqtt/vernemq] Readme 

### PR DESCRIPTION
**Pull request by mprevide**
* **Please check if the PR fulfills these requirements**
  * [ ] Tests for the changes have been added (for bug fixes / features)
  * [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
docs
* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**
Add documentation for mqtt/vernemq 

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)

#1330, #1305, #1304

* **Other information**:

### CODE REVIEW

File: connector/mqtt/vernemq/README.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
+# MQTT 
+
+## Overview
+
+### Iot-Agent MQTT for Dojot
+
+The Dojot VerneMQ service is a extension of [VerneMQ](https://github.com/vernemq/vernemq) with some features for dojot case.
+
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/diag_mqtt)
```
Comment (mprevide) [inline connector/mqtt/vernemq/README.md]: You can see the image before merge here
http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/mprevide/dojot/readme_mqtt/connector/mqtt/vernemq/docs/plant_uml/mqtt/diag_mqtt

File: connector/mqtt/vernemq/README.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+The process of obtaining certificates for client (Fig. 2):
+
+- Creation of entity at EJBCA, usually this unique entity consists of ***tenant***_:_***device_id***.
+
+- Obtaining a public certificate for the client, which happens with the creation of the CSR (Certificate Signing Request, it usually contains the public key, identifying information such as a hostname and unique identification) and its submission to obtain a public certificate.
+
+- Obtaining the root certificate (public certificate of **CA**).
+
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/seq_sec_client)
```
Comment (mprevide) [inline connector/mqtt/vernemq/README.md]: You can see the image before merge here
http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/mprevide/dojot/readme_mqtt/connector/mqtt/vernemq/docs/plant_uml/mqtt/seq_sec_client

File: connector/mqtt/vernemq/README.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+The process of obtaining certificates for VerneMQ, K2V Brige and K2V-bridge (services) instances follows the same steps as for the client, with some more followed by (Fig. 3):
+
+- At each defined time (CHECK_EXPIRATION_TIME), it's checked if the root certificate and public certificate of the service instance will expire in the next CHECKEND_EXPIRATION_SEC seconds.
+
+- At each defined time (CHECK_BROKER_CERT_REVOKED_TIME), it's checked if the public certificate of the service instance has been revoked.
+
+- CRL updated the CRL certificate every time by setting in CRL_UPDATE_TIME
+
+
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/seq_sec_service1)
```
Comment (mprevide) [inline connector/mqtt/vernemq/README.md]: You can look the image before merge here
http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/mprevide/dojot/readme_mqtt/connector/mqtt/vernemq/docs/plant_uml/mqtt/seq_sec_service1

**Comment by mprevide** [comment]
Codecov verification is expected, but there are no changes to trigger any coverage.

File: connector/mqtt/vernemq/README.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+Fig. 3 - VerneMQ (Broker) retrive certificates from PKI (EJBCA)
+
+The TLS connection has a maximum life time, see more about [Disconnect Plugin](vernemq/plugins/dojot_disconnect_plugin) for VerneMQ.
+The TLS connection also has a configurable timeout, which is a VerneMQ configuration..
+
+Environment variables mentioned above are more described in [here](./vernemq)
+
+## How to get a certificate for a client to use with Dojot VerneMQ
+
+### **Before**
```
Comment (raulnegreiros) [inline connector/mqtt/vernemq/README.md]: Maybe "Prerequisites" sounds better

File: connector/mqtt/vernemq/README.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
+# MQTT
```
[the remark this answers is not in the document]
Comment (raulnegreiros) [reply connector/mqtt/vernemq/README.md]: I think is not a good idea to put communication between internal services in this document.

File: connector/mqtt/vernemq/README.md [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+
+## Security
+
+We use **TLS** mutual authentication to provide a secure communication between devices and broker [VerneMQ](https://github.com/vernemq/vernemq) through **MQTT**.  Transport Layer Security (**TLS**) is a cryptographic protocol designed to provide communications security over a computer network.
+
+In **TLS** each device and broker should be provisioned with a private key and corresponding public certificate sign from **CA** (certificate authority) and a root certificate (public certificate of **CA**), the **CA** is include in **PKI** (public key infrastructure). A **PKI**  ([EJBCA](./ejbca)) is a system for the creation, storage, and distribution of digital certificates.
+
+Also a PKI includes the certificate revocation list (**CRL**), which is a list of certificates that have been revoked before reaching the expiration date of the certificate.
+
+The process of obtaining certificates for client (Fig. 2):
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: The process of obtaining certificates for a client

File: connector/mqtt/vernemq/README.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/seq_sec_service1)
+
+Fig. 3 - VerneMQ (Broker) retrieves certificates from PKI (EJBCA)
+
+The TLS connection has a maximum life time, see more about [Disconnect Plugin](broker/src/dojot_disconnect_plugin) for VerneMQ.
+The TLS connection also has a configurable timeout, which is a VerneMQ configuration.
+
+Environment variables mentioned above are more described in [here](./vernemq)
+
+## How to connect a device with the IoTAgent-MQTT witn Security (via TLS)
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: How to connect a device with the IoTAgent-MQTT with Security

File: connector/mqtt/vernemq/README.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+#### 3. Create a folder
+
+It's  recommended to create a new folder for certificates.
+
+```console
+mkdir -p certs
+cd certs
+```
+
+#### 4. Generate pair of (public and private) RSA key
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: Generate a pair of ...

File: connector/mqtt/vernemq/README.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+ openssl genrsa -out client.key 2048
+```
+
+#### 5. Create entity in ejbca
+
+NOTE: Where ***tenant*** is the information context separation of dojot and ***device_id*** is a unique identification for the device.
+
+##### Create environment variables with device unique identification
+
+Change the value of TENANT and DEVICE_ID variables from block of code below and run:
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: from the block ...

File: connector/mqtt/vernemq/README.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+        --subj "/CN=${TENANT}:${DEVICE_ID}"
+```
+
+And create a environment variable with CSR content:
+
+```console
+export CSR_CONTENT=$(cat client.csr | sed '1,1d;$ d' | tr -d '\r\n')
+```
+
+#### 7. Signing certificate for entity (Get public certificate for the client)
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: Signing certificate for an entity

File: connector/mqtt/vernemq/README.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+(echo  "-----BEGIN CERTIFICATE-----"
+echo ${CRT_ENTITY_CONTENT}
+echo "-----END CERTIFICATE-----" ) > client.crt
+```
+
+#### 8. Obtaining the root certificate (public certificate of **CA**)
+
+##### Retrieve CA Certificate
+
+Retrieve certificate and set in a environment variable
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: Retrieve certificate and set it to an environment variable

File: connector/mqtt/vernemq/README.md [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+echo "-----END CERTIFICATE-----" ) > root.crt
+```
+
+## Simulating a device with mosquitto
+
+NOTE: Ports and address may change depending on the deployment
+
+NOTE 2: The **tenant** is *admin* and **device_id** is *a1998e* for these examples. You must change them for your case.
+
+In these examples we will use [mosquitto](https://mosquitto.org/) client.
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: add a comma after examples.

File: connector/mqtt/vernemq/README.md [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+The tree files are require: *client.crt*, *client.key* and *root.crt*. See example:
+
+##### Example on how to publish
+
+```console
+mosquitto_pub -h myhost -p 8883 -t admin:a1998e/attrs -m '{"attr_example": 10 }' --cert client.crt  --key client.key --cafile root.crt
+```
+
+Note: In this case, the message is a publish  on an attribute with the label attr_example and a new value 10 in device *a1998e* with tenant *admin*.
```
Comment (jonaphael) [inline connector/mqtt/vernemq/README.md]: In this case, the message is a publish on an attribute
Doubt: In this case, the message is a publish with an attribute?

File: connector/mqtt/vernemq/README.md [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+
+```console
+(echo  "-----BEGIN CERTIFICATE-----"
+echo ${CRT_ROOT_CONTENT}
+echo "-----END CERTIFICATE-----" ) > root.crt
+```
+
+## Simulating a device with mosquitto
+
+NOTE: Ports and address may change depending on the deployment
```
Comment (rascaraficci) [inline connector/mqtt/vernemq/README.md]: Port

File: connector/mqtt/vernemq/README.md [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+```console
+export TENANT=admin
+export DEVICE_ID=a1998e
+```
+
+##### Create entity in EJBCA
+
+```console
+curl  -X POST ${DOJOT_URL}/user \
```
[the remark this answers is not in the document]
Comment (mprevide) [reply connector/mqtt/vernemq/README.md]: Yes...

File: connector/mqtt/vernemq/README.md [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+
+##### Address of the dojot
+
+Change the value of the DOJOT_URL variable from the code block below and execute:
+
+```console
+export DOJOT_URL=http://myhost:8000
+```
+
+##### Server hostname
```
[the remark this answers is not in the document]
Comment (mprevide) [reply connector/mqtt/vernemq/README.md]: It's not necessary. I will remove 'Server hostname' =)

File: connector/mqtt/vernemq/README.md [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+
+The IoT-Agent MQTT extends [VerneMQ](https://github.com/vernemq/vernemq) with some features and services for dojot case. Basically, devices publishes MQTT messages in specific topics in VerneMQ. These messages are consumed by V2K-bridge service, which adapt them to the dojot's data model and forwards the modified messages to specific topics in Apache Kafka so that they can be consumed by other services. There is a reverse flow, where services publish messages to Apache Kafka in specific topics, these messages are consumed by the service K2V-bridge which adapts the messages and forwards the modified messages to VerneMQ so that they can be consumed by the devices. These data flows are depicted in Fig. 1.
+
+The currently accepted **MQTT protocol versions** are MQTT v3.1 and v3.1.1 respectively.
+
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/diag_mqtt_1)
+
```
Comment (mprevide) [inline connector/mqtt/vernemq/README.md]: You can look at the image before merging here
http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/mprevide/dojot/readme_mqtt/connector/mqtt/vernemq/docs/plant_uml/mqtt/diag_mqtt_1

File: connector/mqtt/vernemq/README.md [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+- For publish: ***tenant***_:_***device_id***_/attrs_
+- For subscription: ***tenant***_:_***device_id***_/config_
+
+Where ***tenant*** is the information context separation of dojot and ***device_id*** is a unique identification for the device.
+The junction (***tenant:device_id***) of *tenant* and  *device_id* always be unique.
+
+See more about [ACL Plugin](broker/src/dojot_acl_plugin) for VerneMQ.
+
+## Security
```
[the remark this answers is not in the document]
Comment (mprevide) [reply connector/mqtt/vernemq/README.md]: This topic will be updated soon. We are developing a new x509 identity solution at dojot. =)

File: connector/mqtt/vernemq/README.md [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+
+## Overview
+
+An IoT-Agent is an adaptation service between physical devices and the dojot platform. The IoT-Agents are responsible for receiving messages from physical devices (directly or through a gateway) and sending them commands in order to configure. The dojot platform can have multiple IoT-Agents, each one of them being specialized in a specific protocol like in this case MQTT. It is also responsible for ensuring secure communication with devices.
+
+The IoT-Agent MQTT extends [VerneMQ](https://github.com/vernemq/vernemq) with some features and services for dojot case. Basically, devices publishes MQTT messages in specific topics in VerneMQ. These messages are consumed by V2K-bridge service, which adapt them to the dojot's data model and forwards the modified messages to specific topics in Apache Kafka so that they can be consumed by other services. There is a reverse flow, where services publish messages to Apache Kafka in specific topics, these messages are consumed by the service K2V-bridge which adapts the messages and forwards the modified messages to VerneMQ so that they can be consumed by the devices. These data flows are depicted in Fig. 1.
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: "[...] by the K2V-bridge service, [...]"

File: connector/mqtt/vernemq/README.md [hunk 20]
```diff
@@ -0,0 +1,306 @@
...
+
+The currently accepted **MQTT protocol versions** are MQTT v3.1 and v3.1.1 respectively.
+
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/diag_mqtt_1)
+
+Fig. 1 - Data communication flows among the services that implement the IoT-Agent MQTT.
+
+### VerneMQ Broker with Custom to Dojot
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: I do not understand this title, the phraseology looks awkward.

File: connector/mqtt/vernemq/README.md [hunk 21]
```diff
@@ -0,0 +1,306 @@
...
+
+The VerneMQ Broker with Custom to Dojot has some bash scripts were added to integrate VerneMQ with dojot's PKI. These scripts are responsible for providing an x509 certificate to the broker and periodically obtaining the CRL with revoked certificates.
+
+For more details, please check the documentation at [the service repository](./broker)
+
+### V2K-bridge
+
+The V2K-bridge service implements a bridge between VerneMQ broker and Kafka broker. Basically, it subscribes to some MQTT topics and forwards the messages to some Kafka topics following the dojot's topics rules. In order to scale the bridge service, shared MQTT subscriptions are used, which allows to instantiating a group of consumers, i.e. a group of bridge instances, when necessary. The communication between the bridge and the VerneMQ is secured with mutual TLS. Soon, communication with Kafka will also use mutual TLS.
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: "[...] which allows the instantiation of a group of consumers [...]"

File: connector/mqtt/vernemq/README.md [hunk 22]
```diff
@@ -0,0 +1,306 @@
...
+For more see [here](./k2v-bridge)
+
+## ACL (Access-Control List)
+
+An **ACL** (Access-Control List) based authorization is provided to manage permissions so that a device only publish and subscribe to its own topics, which are:
+
+- To publish: ***tenant***_:_***device_id***_/attrs_
+- To subscribe: ***tenant***_:_***device_id***_/config_
+
+Where ***tenant*** is a context identifier into dojot and ***device_id*** is a identifier for the device in the corresponding context. Joining both identifiers (***tenant:device_id***), you have a unique identifier for the device into dojot.
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: "[...]. By joining both identifiers [...]"

File: connector/mqtt/vernemq/README.md [hunk 23]
```diff
@@ -0,0 +1,306 @@
...
+
+See more about [ACL Plugin](./broker/src/dojot_acl_plugin) for VerneMQ.
+
+## Security
+
+**Note: This topic will be updated soon. We are developing a new x509 identity solution at dojot.**
+
+We use **TLS** mutual authentication to provide secure communication between devices and broker [VerneMQ](https://github.com/vernemq/vernemq) through **MQTT**.  Transport Layer Security (**TLS**) is a cryptographic protocol designed to provide communications security over a computer network.
+
+In **TLS** each device and broker should be provisioned with a private key and corresponding public certificate sign from **CA** (certificate authority) and a root certificate (public certificate of **CA**), the **CA** is include in **PKI** (public key infrastructure). A **PKI**  ([EJBCA](./pki/ejbca)) is a system for the creation, storage, and distribution of digital certificates.
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: "[...] the **CA** is included in **PKI** [...]"

File: connector/mqtt/vernemq/README.md [hunk 24]
```diff
@@ -0,0 +1,306 @@
...
+- CRL updated the CRL certificate every time by setting in CRL_UPDATE_TIME
+
+![image](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/dojot/dojot/epic-100kMqttDevices/connector/mqtt/vernemq/docs/plant_uml/mqtt/seq_sec_service1)
+
+Fig. 3 - VerneMQ (Broker) retrieves certificates from PKI (EJBCA)
+
+The TLS connection has a maximum life time, see more about [Disconnect Plugin](./broker/src/dojot_disconnect_plugin) for VerneMQ.
+The TLS connection also has a configurable timeout, which is a VerneMQ configuration.
+
+Environment variables mentioned above are more described in [here](./broker)
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: "[...] above are described in greater detail in [...]"

File: connector/mqtt/vernemq/README.md [hunk 25]
```diff
@@ -0,0 +1,306 @@
...
+NOTE: Ports and addresses may change depending on the deployment
+
+Environment variables
+
+### Steps
+
+
+#### 1. Set variables environment
+
+##### Address of the dojot
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: Dojot's address

File: connector/mqtt/vernemq/README.md [hunk 26]
```diff
@@ -0,0 +1,306 @@
...
+
+Create a *client.key* file with a pair of keys.
+
+```console
+ openssl genrsa -out client.key 2048
+```
+
+#### 5. Create entity in ejbca
+
+##### Create environment variables with device unique identification
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: "[...] unique device identification [...]"

File: connector/mqtt/vernemq/README.md [hunk 27]
```diff
@@ -0,0 +1,306 @@
...
+
+- **--key**: Define the path to a file containing a PEM encoded private key for this client, if required by the server. In this case: *client.key*
+
+- **-m**: Send a single message from the command line.
+
+- **-u**: Provide a username to be used for authenticating with the broker.
+
+### With security: use TLS to communicate with VerneMQ
+
+The tree files are require: *client.crt*, *client.key* and *root.crt*. See example:
```
Comment (eduardogmisiuk) [inline connector/mqtt/vernemq/README.md]: three

**Pull request merged**